### PR TITLE
[RLlib] New ConnectorV3 API #05: PPO runs in single-agent mode in this API stack

### PR DESCRIPTION
**Pull request by sven1977**
EnvRunners support new ConnectorV3 API; PPO runs in single-agent mode in this API stack
This PR:
* Adds a new config key: `train_batch_size_per_learner` to better distinguish between total effective batch size and batch size per (GPU) learner worker.
* Makes large changes to the PPO algorithm when run with the new API stack + EnvRunners:
* * Forwards episode data directly from EnvRunner(s) to Learner worker(s) w/o having to form a MultiAgentBatch first.
* * Removes need for PPO's `forward_exploration` to perform a value-function pass. This is an essential improvement in code quality as we now have full separation between the sampling- and the learning worlds. The EnvRunner (sampling world) is no longer concerned with having to think about what the PPOLearner (learning world) might need and only needs to compute actions for the next env step.
* * All vf-computations, GAE, and advantages computations have been moved to the Learner side and these operations are now performed in a batched fashion (on all provided episodes at once). Having the episodes still intact on the Learner side helps reducing the complexity of these computations.

Benchmark results:
Learns Pong in ~5min via examples/connectors/connector_v2_frame_stacking.py example script:

Args: `--num-gpus=8 --num-env-runners=95 --framework=torch`

on commit: 790a537079f998ffc676f8c6dd9b9e3d1b26524d
```
Trial status: 1 RUNNING
Current time: 2024-01-12 12:41:55. Total running time: 7min 0s
Logical resource usage: 96.0/96 CPUs, 8.0/8 GPUs (0.0/1.0 accelerator_type:V100)
╭──────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────╮
│ Trial name            status       iter     total time (s)       ts     reward     episode_reward_max     episode_reward_min     episode_len_mean     episodes_this_iter │
├──────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┤
│ PPO_env_0b2b7_00000   RUNNING       226             362.71   904000      19.62                     21                      9              1728.96                      0 │
╰──────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────╯
```

## Checks

- [x] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a 
           method in Tune, I've added it in `doc/source/tune/api/` under the 
           corresponding `.rst` file.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: rllib/algorithms/ppo/ppo.py [hunk 1]
```diff
@@ -376,57 +402,122 @@ def get_default_policy_class(
             return PPOTF2Policy
 
     @override(Algorithm)
-    def training_step(self) -> ResultDict:
-        use_rollout_worker = self.config.env_runner_cls is None or issubclass(
-            self.config.env_runner_cls, RolloutWorker
-        )
+    def training_step(self):
```
Comment (sven1977) [inline rllib/algorithms/ppo/ppo.py:405]: Disseminate these two stacks now for better readability (avoid too many if-blocks inside `training_step`).

Old- and hybrid stacks will be deprecated soon anyways.

File: rllib/algorithms/ppo/ppo.py [hunk 2]
```diff
@@ -550,24 +638,3 @@ def training_step(self) -> ResultDict:
         self.workers.local_worker().set_global_vars(global_vars)
 
         return train_results
-
-    def postprocess_episodes(
```
Comment (sven1977) [inline rllib/algorithms/ppo/ppo.py:554]: No longer needed here. Episodes are sent directly to Learner(s) as-is.

File: rllib/algorithms/ppo/ppo_learner.py [hunk 3]
```diff
@@ -39,6 +47,78 @@ def build(self) -> None:
             )
         )
 
+    @override(Learner)
+    def _preprocess_train_data(
```
Comment (sven1977) [inline rllib/algorithms/ppo/ppo_learner.py:51]: Note: Only called on the new API stack + EnvRunners.

File: rllib/algorithms/ppo/ppo_learner.py [hunk 4]
```diff
@@ -39,6 +47,78 @@ def build(self) -> None:
...
+        self,
+        *,
+        batch,
+        episodes,
+    ):
+        batch = batch or {}
+        if not episodes:
+            return batch, episodes
+
+        # Make all episodes one ts longer in order to just have a single batch
```
Comment (sven1977) [inline rllib/algorithms/ppo/ppo_learner.py:61]: New way to do GAE:
* elongate all episodes by one artificial ts.
* perform vf-predictions AND bootstrap value predictions in one single batch (b/c we have the extra timestep!)
* * use the learner connector to make sure this forward pass is done using the correct (custom?) batch format.
* remove extra timesteps from episodes (and computed advantages)

File: rllib/algorithms/ppo/ppo_rl_module.py [hunk 5]
```diff
@@ -45,22 +45,19 @@ def get_initial_state(self) -> dict:
...
-        return [SampleBatch.OBS]
+        return self.input_specs_inference()
 
     @override(RLModule)
     def output_specs_exploration(self) -> SpecDict:
-        return [
-            SampleBatch.VF_PREDS,
-            SampleBatch.ACTION_DIST_INPUTS,
-        ]
+        return self.output_specs_inference()
```
Comment (sven1977) [inline rllib/algorithms/ppo/ppo_rl_module.py:60]: simplified

File: rllib/algorithms/ppo/tf/ppo_tf_rl_module.py [hunk 6]
```diff
@@ -40,6 +40,11 @@ def _forward_exploration(self, batch: NestedDict) -> Dict[str, Any]:
         the policy distribution to be used for computing KL divergence between the old
         policy and the new policy during training.
         """
+        # TODO (sven): Make this the only bahevior once PPO has been migrated
+        #  to new API stack (including EnvRunners!).
+        if self.config.model_config_dict.get("uses_new_env_runners"):
```
Comment (sven1977) [inline rllib/algorithms/ppo/tf/ppo_tf_rl_module.py:45]: temporary hack to make sure RLModule knows, when it still has to compute vf-preds via `forward_exploration` (old and hybrid API stacks).

File: rllib/core/learner/learner.py [hunk 7]
```diff
@@ -272,6 +281,40 @@ def __init__(
         # the final results dict in the `self.compile_update_results()` method.
         self._metrics = defaultdict(dict)
 
+    @OverrideToImplementCustomLogic_CallToSuperRecommended
```
Comment (sven1977) [inline rllib/core/learner/learner.py:284]: Moved here for better ordering of methods (used to be all the way at the bottom of class).

File: rllib/core/learner/learner.py [hunk 8]
```diff
@@ -272,6 +281,40 @@ def __init__(
...
+        This method should be called before the learner is used. It is responsible for
+        setting up the RLModule, optimizers, and (optionally) their lr-schedulers.
+        """
+        if self._is_built:
+            logger.debug("Learner already built. Skipping build.")
+            return
+
+        # Build learner connector pipeline used on this Learner worker.
+        # TODO (sven): Support multi-agent cases.
+        if self.config.uses_new_env_runners and not self.config.is_multi_agent():
```
Comment (sven1977) [inline rllib/core/learner/learner.py:297]: For now: Only on new API stack + EnvRunner + single-agent: use Learner connector (w/o this PPO on new stack would not learn).

File: rllib/utils/minibatch_utils.py [hunk 9]
```diff
@@ -87,7 +86,13 @@ def __iter__(self):
                     def get_len(b):
                         return len(b[SampleBatch.SEQ_LENS])
 
+                    n_steps = int(
```
Comment (sven1977) [inline rllib/utils/minibatch_utils.py:89]: Bug fix. When slicing on a BxT batch, we should slice properly along B-axis (with the correct slice size!).

File: rllib/policy/sample_batch.py [hunk 10]
```diff
@@ -1207,21 +1187,10 @@ def map_(path, value):
...
-                ):
-                    return value[start:stop]
-                else:
-                    # Since infos should be stored as lists and not arrays,
-                    # we return the values here and slice them separately
-                    # TODO(Artur): Clean this hack up.
-                    return value
-
-            data = tree.map_structure(map_, self)
+            infos = self.pop(SampleBatch.INFOS, None)
```
Comment (sven1977) [inline rllib/policy/sample_batch.py:1190]: Simplifications.

File: rllib/policy/sample_batch.py [hunk 11]
```diff
@@ -1169,34 +1169,14 @@ def map_(path, value):
...
-                            )
-                            or (tf and tf.is_tensor(value) and tf.size(value) > 0)
-                        ):
-                            return value[start_unpadded:stop_unpadded]
-                        else:
-                            # Since infos should be stored as lists and not arrays,
-                            # we return the values here and slice them separately
-                            # TODO(Artur): Clean this hack up.
-                            return value
+                    return value[start_padded:stop_padded]
```
Comment (sven1977) [inline rllib/policy/sample_batch.py:1172]: Simplifications.

**Review by kouroshHakha — APPROVED** [review]
Stmp

**Comment by Mark2000** [comment]
@sven1977 Could you speak more to why GAE support was dropped for APPO in this release?

```
        # IMPALA and APPO need vtrace (A3C Policies no longer exist).
         if not self.vtrace:
             raise ValueError(
                 "IMPALA and APPO do NOT support vtrace=False anymore! Set "
                 "`config.training(vtrace=True)`."
             )
```

**Pull request merged**